core/src/main/scala/Block.scala

```diff
@@ -45,4 +45,25 @@ class Block(val records: LazyList[Record]) extends AnyVal {
       // TODO: Handle exceptions
     } finally writer.close()
   }
+
+  def partition(workers: List[WorkerMetadata]): List[Partition] = {
+    def isInKeyRange(key: Key, range: KeyRange): Boolean =
+      (key >= range._1) && (key <= range._2)
+
+    val groupedRecords = for {
+      worker <- workers
+      keyRange = worker.keyRange.getOrElse(
+        throw new AssertionError("KeyRange must be defined")
+      )
+      filteredRecords = records.filter(record =>
+        isInKeyRange(record.key, keyRange)
+      )
+    } yield new Partition(
+      worker,
+      new Block(filteredRecords)
+    )
+
+    groupedRecords
+  }
```

Comment:
Let's refactor in next iteration. https://github.com/users/betarixm/projects/2/views/1?pane=issue&itemId=44045372